stream-chat-android-client/src/test/java/io/getstream/chat/android/client/uploader/StreamFileUploaderTest.kt

```diff
@@ -0,0 +1,222 @@
+package io.getstream.chat.android.client.uploader
+
+import android.webkit.MimeTypeMap
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.eq
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.whenever
+import io.getstream.chat.android.client.api.RetrofitCdnApi
+import io.getstream.chat.android.client.api.models.CompletableResponse
+import io.getstream.chat.android.client.api.models.UploadFileResponse
+import io.getstream.chat.android.client.errors.ChatError
+import io.getstream.chat.android.client.utils.ProgressCallback
+import io.getstream.chat.android.client.utils.RetroError
+import io.getstream.chat.android.client.utils.RetroSuccess
+import org.amshove.kluent.shouldBeEqualTo
+import org.amshove.kluent.shouldBeNull
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.Shadows.shadowOf
+import java.io.File
+
+@RunWith(AndroidJUnit4::class)
+internal class StreamFileUploaderTest {
+
+    private val channelType = "channelType"
+    private val channelId = "channelId"
+    private val userId = "userId"
+    private val connectionId = "connectionId"
+    private val apiKey = "apiKey"
+    private val url = "url"
+
+    private val progressCallback = object : ProgressCallback {
+        override fun onSuccess(file: String) = Unit
+
+        override fun onError(error: ChatError) = Unit
+
+        override fun onProgress(progress: Long) = Unit
+    }
+
+    private lateinit var retrofitCdnApi: RetrofitCdnApi
+    private lateinit var streamFileUploader: StreamFileUploader
+
+    @Before
+    fun before() {
+        retrofitCdnApi = mock()
+        shadowOf(MimeTypeMap.getSingleton())
+        streamFileUploader = StreamFileUploader(apiKey, retrofitCdnApi)
+    }
+
+    @Test
+    fun `Should send file to api when sending file without progress callback`() {
+        whenever(retrofitCdnApi.sendFile(any(), any(), any(), any(), any(), any())).thenReturn(
+            RetroSuccess(UploadFileResponse("file")).toRetrofitCall()
+        )
+
+        streamFileUploader.sendFile(channelType, channelId, userId, connectionId, File(""))
+
+        verify(retrofitCdnApi).sendFile(
+            eq(channelType),
+            eq(channelId),
+            file = any(),
+            eq(apiKey),
+            eq(userId),
+            eq(connectionId)
```

Comment:
I have mixed feelings about these tests. They mostly just copy the implementation of the `StreamFileUploader` class, which is just to pass all parameters forward to the CDN API. Doesn't seem too meaningful. And for the parameter that involves business logic in our own class, creating the form data from the `File`, we aren't asserting anything.

The question I always ask: _What will it mean if this unit test fails?_ If it just means that we've updated the implementation, so now we have to come to this other file and update it the same way... That doesn't sound very productive.

Reply:
Good point. Most of them are checking if we call CDN API with proper parameters and I agree that if something change, we will need to change the code in both places. Let me improve that on Monday